### ISSUE TITLE
Generative Contrastive Graph Learning for Recommendation

### ISSUE THREAD

**Issue by wonsutak-ysz** [AIML-K/GNN_Survey#37]
```bib
@misc{yang2023generativecontrastivegraphlearning,
      title={Generative Contrastive Graph Learning for Recommendation}, 
      author={Yonghui Yang and Zhengwei Wu and Le Wu and Kun Zhang and Richang Hong and Zhiqiang Zhang and Jun Zhou and Meng Wang},
      year={2023},
      eprint={2307.05100},
      archivePrefix={arXiv},
      primaryClass={cs.IR},
      url={https://arxiv.org/abs/2307.05100}, 
}
```

**Comment by wonsutak-ysz** [comment]
### introduction
Graph learning models, which view user interactions as user-item graphs, have been widely applied in collaborative filtering-based recommendations. Recently, researchers have introduced graph contrastive learning techniques into collaborative filtering to mitigate the sparse supervision problem, first constructing contrastive views through data augmentation, then providing self-supervised signals by maximizing mutual information between contrastive views. Despite their effectiveness, current graph contrastive learning-based recommendation models still have limitations compared to existing data augmentation techniques, both in structural and feature augmentation. First, structural augmentation randomly removes nodes or edges, which can easily destroy the intrinsic properties of user-item graphs; second, feature augmentation applies the same scale of noise to each node, ignoring the unique characteristics of nodes in the graph. To address these limitations, this paper proposes a new variational graph generative contrastive learning framework for recommendations. It uses variational graph reconstruction to estimate Gaussian distributions for each node, then performs multiple samplings from the estimated distributions to generate multiple contrastive views, establishing a bridge between generative learning and contrastive learning. The generated contrastive views can well reconstruct the input graph without information distortion. Furthermore, it customizes estimated variances for each node to regulate the scale of contrastive loss during optimization. Considering the similarity of estimated distributions, a cluster-aware dual contrastive learning approach is proposed, where node-level contrasting encourages consistency between node contrastive views, and cluster-level contrasting encourages consistency among intra-cluster nodes. Finally, extensive experimental results on three public datasets demonstrate the effectiveness of the proposed model.

**Comment by wonsutak-ysz** [comment]
# Method
![image](https://github.com/user-attachments/assets/23ccbdc8-1286-412c-bbb7-7b64c4dbf6ee)

## Variational graph reconstruction
### Graph Inference

Given a user-item interaction graph $G = {U \cup V, \mathbf{A}}$, with nodes initialized into $\mathbf{E}^0$. Graph inference aims to learn the probability distribution $\mathbf{Z}$, and reconstruct the input graph structure $\hat{\mathbf{A}} \sim P_\theta(\mathbf{A}|\mathbf{Z})$. Similar to VAE models, variational inference is used to approximate the posterior probability $p_\theta(\mathbf{A}|\mathbf{Z})$. Specifically, each node is encoded into a multivariate Gaussian distribution:

$q_\phi(z_i|\mathbf{A}, \mathbf{E}^0) = \mathcal{N}(z_i|\mu_\phi(i), \text{diag}(\sigma_\phi^2(i)))$

where $\mu_\phi(i)$ and $\sigma_\phi^2(i)$ represent the mean and variance of the node's distribution respectively. To better explore high-order user-item graph structures, GNN is used to estimate the parameters of node distributions:

$\mu = GNN(\mathbf{A}, \mathbf{E}^0, \phi_\mu), \sigma = GNN(\mathbf{A}, \mathbf{E}^0, \phi_\sigma)$

where $\phi_\mu$ and $\phi_\sigma$ represent learnable parameters in the graph inference process. This paper uses LightGCN as the encoder to implement the above graph inference process. For each node, its corresponding mean value is updated according to the following formula:

$\mu_i^l = \sum_{j\in\mathcal{N}_i} \frac{1}{\sqrt{|\mathcal{N}_i|}\sqrt{|\mathcal{N}_j|}} \mu_i^{l-1}$

The mean is initialized as $\mu^0 = \mathbf{E}^0$. After L layers of graph convolution, we get the outputs $[\mu^0, \mu^1, \dots, \mu^L]$, then combine all layer outputs and calculate the mean and variance using the following formula:

$\mu = \frac{1}{L}\sum_{l=1}^L \mu^l, \sigma = MLP(\mu)$

where the variance is learned through MLP with mean values as input. In practice, it was found that a single-layer MLP achieved the best performance, and the variance is $\sigma = \exp(\mu\mathbf{W} + \mathbf{b})$, where $\mathbf{W}$ and $\mathbf{b}$ are learnable parameters. After obtaining the approximate posterior's mean and variance, hidden representation $\mathbf{z}_i$ is generated by sampling from distribution $\mathcal{N}(\mu_i, \sigma_i^2)$.

However, since the sampling process is non-differentiable, it cannot be optimized directly. The reparameterization trick is used to replace the sampling process:

$\mathbf{z}_i = \mu_i + \sigma_i \cdot \varepsilon$

where $\varepsilon \sim \mathcal{N}(0, \mathbf{I})$ is standard Gaussian noise.

### Graph Generation
After estimating the probability distribution of latent variable Z, the goal of graph generation is to reconstruct the original user-item graph:
$p(\mathbf{A}|\mathbf{Z}) = \prod_{i=0}^{M+N-1} \prod_{j=0}^{M+N-1} p(\mathbf{A}_{ij}|\mathbf{z}_i, \mathbf{z}_j)$
There are many methods to implement the graph generation process, such as inner product, matrix factorization, and neural networks. This paper uses inner product to calculate the propensity score between nodes i and j:
$p(\mathbf{A}_{ij} = 1|\mathbf{z}_i, \mathbf{z}_j) = \sigma(\mathbf{z}_i^T \mathbf{z}_j)$
where $\sigma(\cdot)$ is the sigmoid function.

**Comment by wonsutak-ysz** [comment]
## Cluster-Aware Contrastive Learning

### Construction of Contrastive Views

Given the estimated latent variable $\mathbf{Z} \sim \mathcal{N}(\mu, \sigma^2)$ probability distribution, we introduce a contrastive learning paradigm based on the estimated distribution. Unlike previous graph-based contrastive learning recommendation methods, we construct contrastive views by sampling multiple times from the estimated distribution, which differs from data augmentation. Specifically, for each node i, the node representations used for contrast are generated according to the following formula:
<img width="130" alt="image" src="https://github.com/user-attachments/assets/0d58959b-5773-4383-a085-a0175ee93ecd">
where $\varepsilon', \varepsilon'' \sim \mathcal{N}(0, \mathbf{I})$ are two random Gaussian noises. Compared to structural augmentation or feature augmentation, this method is more efficient in constructing contrastive views. First, sampling all contrastive representations from the estimated distribution can well reconstruct the input graph without any information distortion. Second, customizing the estimated variance for each node allows adaptive adjustment of the scale of contrastive loss.

### Node-Level Contrastive Loss

After constructing contrastive views for each node, we maximize mutual information to provide self-supervised signals to improve recommendation performance. Considering that similar nodes' representations will be closer, this paper proposes cluster-aware dual contrastive loss functions for optimization: node-level contrastive loss and cluster-level contrastive loss. The node-level contrastive loss encourages consistency between contrastive views of each node, while the cluster-level contrastive loss encourages consistency between contrastive views of nodes within the same cluster.
The node-level contrastive loss is $\mathcal{L}_N = \mathcal{L}_N^U + \mathcal{L}_N^V$, where $\mathcal{L}_N^U$ and $\mathcal{L}_N^V$ represent the losses for users and items respectively:
<img width="295" alt="image" src="https://github.com/user-attachments/assets/6610a80f-f73a-4f90-9084-e60bafceaf1e">

### Cluster-Level Contrastive Loss

Considering the similarity between node estimated distributions, we design cluster-level contrastive loss to further distinguish positive and negative pairs in the same batch of training data. Generally speaking, it aims to maximize the consistency between nodes within the same cluster while minimizing the consistency between nodes from different clusters. Assume there are $K_u$ user cluster prototypes (samples) $\mathbf{C}^u \in \mathbb{R}^{d\times K_u}$ and $K_i$ item cluster prototypes (samples) $\mathbf{C}^i \in \mathbb{R}^{d\times K_i}$. Use conditional probability $p(\mathbf{c}_k^u|\mathbf{z}_a)$ to represent that user $a$ belongs to the $k^{th}$ user class, and conditional probability $p(\mathbf{c}_h^i|\mathbf{z}_i)$ to represent that item $i$ belongs to the $h^{th}$ item class. Given the estimated distribution as input, use K-means algorithm to implement the clustering process. Then calculate the probability of two users/items being assigned to the same prototype:
<img width="247" alt="image" src="https://github.com/user-attachments/assets/39733a76-0f8b-4197-87e7-8d7b640dfa07">
where $p(a,b)$ represents the probability that users $a$ and $b$ belong to the same class, $p(i,j)$ represents the probability that items $i$ and $j$ belong to the same class. Define the cluster-level loss as $\mathcal{L}_C = \mathcal{L}_C^U + \mathcal{L}_C^I$, where $\mathcal{L}_C^U$ and $\mathcal{L}_C^I$ represent the losses for users and items respectively:
<img width="442" alt="image" src="https://github.com/user-attachments/assets/2278c4c5-709e-455d-95f4-97ca98351fcc">
where $SP(a) = \sum_{b\in\mathcal{B}u,b!=a} p(a,b)$, $SP(i) = \sum{j\in\mathcal{B}_i,j!=i} p(i,j)$. $\tau_2$ represents the temperature parameter used to control the scale of minimizing difficult negative samples. The final contrastive loss is the weighted sum of node-level loss and cluster-level loss:
<img width="148" alt="image" src="https://github.com/user-attachments/assets/67bce344-2386-4b7a-865c-33156e8673dc">

### Model Optimization

For the variational graph reconstruction part, ELBO is used to optimize the parameters of graph inference and generation processes:
<img width="486" alt="image" src="https://github.com/user-attachments/assets/ec069cde-eb18-4804-8fd8-1aca6509d3d2">
The first term in the above equation represents the reconstruction error between the original graph and the generated graph. Using pairwise learning strategy to minimize the reconstruction error:
<img width="439" alt="image" src="https://github.com/user-attachments/assets/8f8925d1-2e6c-46cd-a274-22a1f9faadb8">
where $\mathcal{D}_a = {(i,j)|i \in R_a \land j \notin R_a}$ represents user a's pairwise training data. $R_a$ represents the set of items that user a has interacted with. In general, use multi-task learning framework to optimize the proposed VGCL model:
<img width="259" alt="image" src="https://github.com/user-attachments/assets/5723a6cb-e061-4944-88a5-0f3e0ab85c75">
After model training, use formula $p(\mathbf{A}_{ij}=1|\mathbf{z}_i,\mathbf{z}_j) = \sigma(\mathbf{z}_i^T\mathbf{z}_j)$ to predict unknown preferences.

**Comment by wonsutak-ysz** [comment]
# Experiment
## Dataset
![image](https://github.com/user-attachments/assets/95357c4b-42fa-4bd0-8b23-2dad05453362)
## performance comparison
![image](https://github.com/user-attachments/assets/f206c57f-8138-48ad-bce1-478d1c45c75b)
## Ablation experiment
![image](https://github.com/user-attachments/assets/e8b404c3-50df-49a3-ac85-a481130099ea)

To investigate the effectiveness of each component of the proposed VGCL, ablation studies were performed on three datasets. As shown in Table 3. VGCL-w/o C represents removal of cluster-level losses, VGCL-w/o V represents removal of node-level losses, and uses the same feature enhancement method as SimGCL to generate contrast views. Please refer to the original article for more details.
## Discussion of estimated distributions
![image](https://github.com/user-attachments/assets/3be04a53-6b26-44c3-bf60-02dce4c13195)

The VGCL adaptively learns the variance of various nodes. In order to study the effect of personalization variance, different user groups were compared. Specifically, users are first divided into four groups according to their interaction behaviors, and then the recommendation performance under different user groups is analyzed. Figure 3 shows the NDCG@20 values for each group on the Douban-Book and Dianping datasets. Please refer to the original article for more details.

## Hyperparameter sensitivity analysis
![image](https://github.com/user-attachments/assets/0ee6ac19-7026-4ce5-acd5-7922cc55c0f9)

The influence of hyperparameters in VGCL is analyzed, including temperature coefficient in contrast loss, number of centroid of cluster, number of layers in graph inference and weight in loss function. The results are shown below, please refer to the original article for more details.

**Comment by wonsutak-ysz** [comment]
# Conclusion
This paper investigates graph contrastive learning-based recommendation from the perspective of better contrastive view construction and proposes a novel variational graph generative contrastive learning framework. It utilizes variational graph reconstruction techniques to generate contrastive views (not data augmentation) for contrastive learning. First, it estimates the probability distribution of each node through graph variational inference, then generates contrastive views with multiple samples from the estimated distributions. Thus, it establishes a bridge between generative learning models and contrastive learning models for recommendation. There are two main advantages:

The generated contrastive representations can well reconstruct the original graph without information distortion
The estimated variances vary across nodes, allowing adaptive adjustment of the contrastive loss scale for each node

Furthermore, considering the similarity of node estimated distributions, a cluster-aware dual contrastive learning approach is proposed, where node-level contrasting encourages consistency between node contrastive views, and cluster-level contrasting encourages consistency among intra-cluster nodes.
Empirical studies on three public datasets demonstrate the effectiveness of the proposed framework.
